src/reagentai/tools/helpers.py

```diff
@@ -1,55 +1,385 @@
+
+        children = tree_dict.get("children", [])
+        for child in children:
+            self._add_effective_size(child)
+
+        if children:
+            tree_dict["eff_height"] = sum(c["eff_height"] for c in children) + self.margin * (
+                len(children) - 1
+            )
+            tree_dict["eff_width"] = (
+                max(c["eff_width"] for c in children) + tree_dict["image"].width + self.margin
+            )
+        else:
+            tree_dict["eff_height"] = tree_dict["image"].height
+            tree_dict["eff_width"] = tree_dict["image"].width + self.margin
+
+    def _add_pos(self, tree_dict: dict, pos: tuple[int, int]):
+        """Recursively adds position information to the tree dictionary."""
+
+        tree_dict["left"] = pos[0]
+        tree_dict["top"] = pos[1]
+        children = tree_dict.get("children")
+        if not children:
+            return
+        mid_y = pos[1] + int(tree_dict["image"].height * 0.5)
+        children_height = sum(c["eff_height"] for c in children) + self.margin * (
+            len(children) - 1
+        )
+        children_leftmost = pos[0] - self.margin - max(c["image"].width for c in children)
+        child_y = mid_y - int(children_height * 0.5)
+        for child in children:
+            y_adjust = int((child["eff_height"] - child["image"].height) * 0.5)
+            self._add_pos(child, (children_leftmost, child_y + y_adjust))
+            child_y += self.margin + child["eff_height"]
+
+    def _make_image(self, tree_dict: dict):
+        """Draws the molecule image and its connections on the canvas."""
+
+        self.image.paste(tree_dict["image"], (tree_dict["left"], tree_dict["top"]))
+        self._draw_number_on_frame(tree_dict)
+        children = tree_dict.get("children")
+        if not children:
+            return
+        children_right = max(c["left"] + c["image"].width for c in children)
+        mid_x = children_right + int(0.5 * (tree_dict["left"] - children_right))
+        mid_y = tree_dict["top"] + int(tree_dict["image"].height * 0.5)
+        self._draw.line((tree_dict["left"], mid_y, mid_x, mid_y), fill="black", width=3)
+        for child in children:
+            self._make_image(child)
+            child_mid_y = child["top"] + int(0.5 * child["image"].height)
+            self._draw.line(
+                (
+                    mid_x,
+                    mid_y,
+                    mid_x,
+                    child_mid_y,
+                    child["left"] + child["image"].width,
+                    child_mid_y,
+                ),
+                fill="black",
+                width=3,
+            )
+        self._draw.ellipse(
+            (mid_x - 8, mid_y - 8, mid_x + 8, mid_y + 8), fill="black", outline="black"
+        )
+
+    def _draw_number_on_frame(self, tree_dict: dict):
+        """Draws the molecule ID as black text on the top-left corner of its frame."""
```

Comment:
Consider adding full docstrings instead of single lines. To keep it consistent with other functions.